frontend/src/stores/useProjectStore.ts

```diff
@@ -224,6 +245,94 @@ const useProjectStore = create<ProjectStore>()(persist((set: SetState<ProjectSto
     project.states = project.states.map((st: AutomataState) => st.id === state.id ? { ...st, ...state } : st)
   })),
 
+  /* Remove a state by id */
+  removeState: (state: AutomataState) => set(produce(({ project }: { project: StoredProject }) => {
+    project.states = project.states.filter((st: AutomataState) => st.id !== state.id)
+  })),
+
+  insertGroup: (createData, isTemplate = false) => {
+    // Check that we are inserting into same project type
+    if (createData.projectType !== get().project.projectType) {
+      return { type: InsertGroupResponseType.FAIL, body: `You cannot insert elements from a ${createData.projectType} project into a ${get().project.projectType} project.` }
+    }
+    // Check that for transitions being inserted, to and from states are also inserted
+    let missingState = false
+    createData.transitions.forEach((transition) => {
+      let toFound = false
+      let fromFound = false
+      createData.states.forEach((state) => {
+        if (state.id === transition.from) {
+          fromFound = true
+        }
+        if (state.id === transition.to) {
+          toFound = true
+        }
+      })
+      // If a transition has to or from missing, flag to return error
+      if (!(toFound && fromFound)) {
+        missingState = true
+      }
+    })
+    if (missingState) {
+      return { type: InsertGroupResponseType.FAIL, body: 'Sorry, there was an error.' }
+    }
```

Comment:
```suggestion
    const stateIDs = new Set(createData.states.map(it => it.id))
    for (const transition of createData.transitions) {
      const hasTo = stateIDs.has(transition.to)
      const hasFrom = stateIDs.has(transition.from)
      // If either `to` or `from` is missing then return an error
      if (!(hasTo && hasFrom)) {
        return { type: InsertGroupResponseType.FAIL, body: 'Sorry, there was an error.' }
      }
    }
```
Using sets is a bit cleaner imo, lemme know what you think though.
I tired to do the same logic so should work the same (if something is fishy plz mention). Main difference is that we return the error inside the loop so that it breaks early if we know something is wrong